homeassistant/components/lgthinq/__init__.py

```diff
@@ -0,0 +1,129 @@
+"""Support for LG ThinQ Connect device."""
+
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Collection
+import logging
+
+from thinqconnect import ThinQApi, ThinQAPIException
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_COUNTRY, Platform
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryError
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import CONF_CONNECT_CLIENT_ID, DEFAULT_COUNTRY
+from .coordinator import DeviceDataUpdateCoordinator, async_setup_device_coordinator
+
+type ThinqConfigEntry = ConfigEntry[dict[str, DeviceDataUpdateCoordinator]]
+
+PLATFORMS = [Platform.SWITCH]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ThinqConfigEntry) -> bool:
+    """Set up an entry."""
+    access_token = entry.data.get(CONF_ACCESS_TOKEN)
+    client_id = entry.data.get(CONF_CONNECT_CLIENT_ID)
+    country_code = entry.data.get(CONF_COUNTRY, DEFAULT_COUNTRY)
+
+    thinq_api = ThinQApi(
+        session=async_get_clientsession(hass),
+        access_token=access_token,
+        country_code=country_code,
+        client_id=client_id,
+    )
+
+    # Setup coordinators and register devices.
+    await async_setup_coordinators(hass, entry, thinq_api)
+
+    # Set up all platforms for this device/entry.
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Clean up devices they are no longer in use.
+    async_cleanup_device_registry(hass, entry)
+
+    return True
+
+
+async def async_setup_coordinators(
+    hass: HomeAssistant,
+    entry: ThinqConfigEntry,
+    thinq_api: ThinQApi,
+) -> None:
+    """Set up coordinators and register devices."""
+    entry.runtime_data = {}
+
+    # Get a device list from the server.
+    try:
+        device_list = await thinq_api.async_get_device_list()
+    except ThinQAPIException as exc:
+        raise ConfigEntryError(exc.message) from exc
```

Comment:
We'd rather raise ConfigEntryNotReady because then it will retry

Reply:
Modified